MatrixKit/Models/Room/MXKRoomDataSource.m

```diff
@@ -167,41 +167,78 @@ @implementation MXKRoomDataSource
         id sessionStateObserver = nil;
         sessionStateObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionStateDidChangeNotification object:mxSession queue:nil usingBlock:^(NSNotification * _Nonnull note) {
             if (mxSession.state >= MXSessionStateStoreDataReady)
             {
                 [[NSNotificationCenter defaultCenter] removeObserver:sessionStateObserver];
-                if (!roomDataSource.room)
-                {
-                    [self finalizeRoomDataSource:roomDataSource onComplete:onComplete];
-                }
+                [self ensureInitialEventExistenceForDataSource:roomDataSource initialEventId:initialEventId andMatrixSession:mxSession onComplete:onComplete];
             }
         }];
     }
 }
 
-+ (void)loadRoomDataSourceWithRoomId:(NSString*)roomId initialEventId:(NSString*)initialEventId andMatrixSession:(MXSession*)mxSession onComplete:(void (^)(id roomDataSource))onComplete
+/// Ensure initial event existence for the roomDataSource.
++ (void)ensureInitialEventExistenceForDataSource:(MXKRoomDataSource*)roomDataSource initialEventId:(NSString*)initialEventId andMatrixSession:(MXSession*)mxSession onComplete:(void (^)(id roomDataSource))onComplete
 {
-    MXKRoomDataSource *roomDataSource = [[self alloc] initWithRoomId:roomId initialEventId:initialEventId andMatrixSession:mxSession];
-    [self finalizeRoomDataSource:roomDataSource onComplete:onComplete];
-}
-
-+ (void)loadRoomDataSourceWithPeekingRoom:(MXPeekingRoom*)peekingRoom andInitialEventId:(NSString*)initialEventId onComplete:(void (^)(id roomDataSource))onComplete
-{
-    MXKRoomDataSource *roomDataSource = [[self alloc] initWithPeekingRoom:peekingRoom andInitialEventId:initialEventId];
-    [self finalizeRoomDataSource:roomDataSource onComplete:onComplete];
+    if (roomDataSource.room)
+    {
+        //  already finalized
+        return;
+    }
+    
+    if (initialEventId == nil)
+    {
+        //  if an initialEventId not provided, finalize
+        [self finalizeRoomDataSource:roomDataSource onComplete:onComplete];
+        return;
+    }
+    
+    //  ensure event with id 'initialEventId' exists in the session store
+    if ([mxSession.store eventExistsWithEventId:initialEventId inRoom:roomDataSource.roomId])
+    {
+        [self finalizeRoomDataSource:roomDataSource onComplete:onComplete];
+    }
+    else
+    {
+        //  wait for the specific event to be existent
+        //  use kMXSessionDidSyncNotification here instead of MXSessionStateRunning, because session does not send this state update if it's already in this state
+        id syncObserver = nil;
+        syncObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionDidSyncNotification object:mxSession queue:nil usingBlock:^(NSNotification * _Nonnull note) {
+            if ([mxSession.store eventExistsWithEventId:initialEventId inRoom:roomDataSource.roomId])
```

Comment:
The event may be never in the store.
If the event is old and the room very active, there will be a gap in the `/sync` that provides only the 10, 20, 30 last messages of a room.
We should give a chance during only one `/sync` and then leave.
If the event is not in the store, the code should be able to open the timeline around this event using `/context` around the event.

Reply:
Ok, updating.